src/tqec/sketchup/block_graph.py

```diff
@@ -562,3 +584,45 @@ def display(
             filepath_or_bytes=bytes_buffer.getvalue(),
             write_html_filepath=write_html_filepath,
         )
+
+    def get_abstract_observables(self) -> list[AbstractObservable]:
+        """Get all the abstract observables from the block graph."""
+        self.check_validity(allow_virtual_node=False)
+        correlation_subgraphs = self.to_zx_graph().find_correlation_subgraphs()
+        abstract_observables: list[AbstractObservable] = []
+
+        def is_measured(cube: Cube) -> bool:
+            return self.get_pipe(cube.position, cube.position.shift_by(0, 0, 1)) is None
+
+        for g in correlation_subgraphs:
+            if g.num_nodes == 1:
+                cube = ty.cast(Cube, self.get_cube(g.nodes[0].position))
+                abstract_observables.append(
+                    AbstractObservable(frozenset({cube}), frozenset())
+                )
+                continue
+            top_lines: set[Cube | Pipe] = set()
+            bottom_regions: set[Pipe] = set()
+            for edge in g.edges:
+                pipe = ty.cast(Pipe, self.get_pipe(edge.u.position, edge.v.position))
```

Comment:
Because having a `None` here should not be possible, I would add an assert instead of a typing cast that has no runtime implication. 
```suggestion
                pipe = self.get_pipe(edge.u.position, edge.v.position)
                assert pipe is not None, f"{edge} is in the graph and should be associated with a Pipe instance."
```